src/services/tiers.service.ts

```diff
@@ -0,0 +1,82 @@
+import { Service, Tier, TiersRepository } from "../core/users/MongoDBTiersRepository";
+import { User } from "../core/users/User";
+import { UsersService } from "./users.service";
+import { createOrUpdateUser, updateUserTier } from './storage.service';
+import { AppConfig } from "../config";
+
+export class TiersService {
+  constructor(
+    private readonly usersService: UsersService,
+    private readonly tiersRepository: TiersRepository,
+    private readonly config: AppConfig
+  ) {}
+
+  async applyTier(userWithEmail: User & { email: string }, productId: string): Promise<void> {
+    const tier = await this.tiersRepository.findByProductId(productId);
+
+    if (!tier) {
+      throw new Error(`Tier for product ${productId} not found`);
```

Comment:
nit: it might be better to throw a custom Error to identify that it is a 404 status error.

Reply:
Done